File: apps/agent/src/agents/workspaceAgent/tools/addCalendarEvent.ts
import { FunctionTool } from '@google/adk';
import type { EventProjection } from '@lifecoach/shared-types';
import { z } from 'zod';
import type { WorkspaceTokensStore } from '../../../storage/workspaceTokens.js';
import type { ExecFileLike, GwsErrorCode } from '../gwsExec.js';
import { projectCalendarEvent } from '../projections/calendarEvent.js';
import { type RunGwsLogEvent, runGws } from '../runGws.js';

/**
 * `add_calendar_event` — single-step calendar.events.insert. Defaults
 * `end` to start + 30 min if omitted; defaults calendarId to "primary".
 */

export const ADD_CALENDAR_EVENT_TOOL_NAME = 'add_calendar_event';
const DEFAULT_DURATION_MS = 30 * 60_000;

export interface CreateAddCalendarEventToolDeps {
  store: WorkspaceTokensStore;
  uid: string;
  execFile?: ExecFileLike;
  log?: (event: RunGwsLogEvent) => void;
}

export type AddCalendarEventResult =
  | { status: 'ok'; event: EventProjection }
  | { status: 'error'; code: GwsErrorCode; message: string };

const parameters = z.object({
  summary: z.string().min(1).describe('Event title — what shows up on the calendar.'),
  start: z
    .string()
    .min(1)
    .describe(
      'RFC3339 start timestamp with timezone offset (e.g. "2026-05-12T18:00:00+01:00"), or YYYY-MM-DD for an all-day event.',
    ),
  end: z
    .string()
    .optional()
    .describe('RFC3339 end (or YYYY-MM-DD all-day). Default = start + 30 minutes.'),
  location: z.string().optional().describe('Optional event location.'),
  description: z.string().optional().describe('Optional event description / notes.'),
  calendarId: z.string().optional().describe('Calendar id. Default "primary".'),
});

interface RawInsertedEvent {
  id?: string;
  summary?: string;
  start?: { dateTime?: string; date?: string; timeZone?: string };
  end?: { dateTime?: string; date?: string; timeZone?: string };
  location?: string;
  htmlLink?: string;
  status?: string;
  description?: string;
}

export function createAddCalendarEventTool(deps: CreateAddCalendarEventToolDeps): FunctionTool {
  const { store, uid, execFile, log } = deps;

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: ADD_CALENDAR_EVENT_TOOL_NAME,
    description:
      "Add a single event to the user's Google Calendar. Use after the user confirms (e.g. via ask_single_choice_question). Returns the created event.",
    parameters,
    execute: async (input: unknown): Promise<AddCalendarEventResult> => {
      const args = input as {
        summary: string;
        start: string;
        end?: string;
        location?: string;
        description?: string;
        calendarId?: string;
      };
      const calendarId = args.calendarId ?? 'primary';
      const isAllDay = /^\d{4}-\d{2}-\d{2}$/.test(args.start);

      const startBlock = isAllDay ? { date: args.start } : { dateTime: args.start };
      const endBlock = args.end
        ? isAllDay
          ? { date: args.end }
          : { dateTime: args.end }
        : isAllDay
          ? { date: args.start }
          : { dateTime: addMs(args.start, DEFAULT_DURATION_MS) };

      const requestBody: Record<string, unknown> = {
        summary: args.summary,
        start: startBlock,
        end: endBlock,
      };
      if (args.location) requestBody.location = args.location;
      if (args.description) requestBody.description = args.description;

      const result = await runGws({
        store,
        uid,
        toolName: ADD_CALENDAR_EVENT_TOOL_NAME,
        service: 'calendar',
        resource: 'events',
        method: 'insert',
        params: { calendarId },
        body: requestBody,
        execFile,
        log,
      });
      if (result.status === 'error') {
        return { status: 'error', code: result.code, message: result.message };
      }
      const projection = projectCalendarEvent(
        ((result.body as RawInsertedEvent | null) ?? {}) as RawInsertedEvent,
        calendarId,
      );
      return { status: 'ok', event: projection };
    },
  });
}

function addMs(rfc3339: string, ms: number): string {
  // Preserve the original timezone offset by manipulating ms only.
  const ts = Date.parse(rfc3339);
  if (Number.isNaN(ts)) return rfc3339;
  const tzMatch = rfc3339.match(/([+-]\d{2}:\d{2}|Z)$/);
  const tz = tzMatch?.[1] ?? 'Z';
  const next = new Date(ts + ms);
  if (tz === 'Z') return next.toISOString();
  // Reconstruct the offset-bearing form by computing the offset in minutes.
  const sign = tz.startsWith('-') ? -1 : 1;
  const parts = tz.replace(/[+-]/, '').split(':');
  const hh = Number(parts[0] ?? '0');
  const mm = Number(parts[1] ?? '0');
  const offsetMinutes = sign * (hh * 60 + mm);
  const local = new Date(next.getTime() + offsetMinutes * 60_000);
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${local.getUTCFullYear()}-${pad(local.getUTCMonth() + 1)}-${pad(local.getUTCDate())}T${pad(local.getUTCHours())}:${pad(local.getUTCMinutes())}:${pad(local.getUTCSeconds())}${tz}`;
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/tools/add_calendar_event.py
"""`add_calendar_event` — single-step calendar.events.insert. Defaults
`end` to start + 30 min if omitted; defaults calendarId to "primary".
"""

from __future__ import annotations

import re
from datetime import UTC, datetime, timedelta
from typing import Any

from lifecoach_agent.workspace_agent.projections import project_calendar_event
from lifecoach_agent.workspace_agent.run_gws import RunGwsOk, run_gws
from lifecoach_agent.workspace_agent.tools._deps import WorkspaceToolDeps

ADD_CALENDAR_EVENT_TOOL_NAME = "add_calendar_event"
_DEFAULT_DURATION_MIN = 30
_ALL_DAY_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def create_add_calendar_event_tool(deps: WorkspaceToolDeps) -> Any:
    async def add_calendar_event(
        summary: str,
        start: str,
        end: str | None = None,
        location: str | None = None,
        description: str | None = None,
        calendarId: str = "primary",  # noqa: N803
    ) -> dict[str, Any]:
        """Add a single event to the user's Google Calendar. Use after
        the user confirms (e.g. via ask_single_choice_question). Returns
        the created event.

        Args:
            summary: Event title — what shows up on the calendar.
            start: RFC3339 start timestamp with timezone offset
                (e.g. "2026-05-12T18:00:00+01:00"), or YYYY-MM-DD for an
                all-day event.
            end: RFC3339 end (or YYYY-MM-DD all-day). Default =
                start + 30 minutes.
            location: Optional event location.
            description: Optional event description / notes.
            calendarId: Calendar id. Default "primary".
        """
        is_all_day = bool(_ALL_DAY_RE.match(start))
        start_block = {"date": start} if is_all_day else {"dateTime": start}

        if end:
            end_block = {"date": end} if is_all_day else {"dateTime": end}
        elif is_all_day:
            end_block = {"date": _add_days_iso(start, 1)}
        else:
            end_block = {"dateTime": _add_minutes(start, _DEFAULT_DURATION_MIN)}

        request_body: dict[str, Any] = {
            "summary": summary,
            "start": start_block,
            "end": end_block,
        }
        if location:
            request_body["location"] = location
        if description:
            request_body["description"] = description

        result = await run_gws(
            store=deps.store,
            uid=deps.uid,
            tool_name=ADD_CALENDAR_EVENT_TOOL_NAME,
            service="calendar",
            resource="events",
            method="insert",
            params={"calendarId": calendarId},
            body=request_body,
            build_client=deps.build_client,
            log=deps.log,
        )
        if not isinstance(result, RunGwsOk):
            return {"status": "error", "code": result.code, "message": result.message}
        raw = result.body if isinstance(result.body, dict) else {}
        projection = project_calendar_event(raw, calendarId)
        return {"status": "ok", "event": projection.model_dump(by_alias=True, exclude_none=True)}

    from google.adk.tools import FunctionTool

    add_calendar_event.__name__ = ADD_CALENDAR_EVENT_TOOL_NAME
    return FunctionTool(add_calendar_event)


def _add_days_iso(yyyy_mm_dd: str, days: int) -> str:
    try:
        d = datetime.fromisoformat(f"{yyyy_mm_dd}T00:00:00+00:00")
    except ValueError:
        return yyyy_mm_dd
    return (d + timedelta(days=days)).date().isoformat()


def _add_minutes(rfc3339: str, minutes: int) -> str:
    try:
        ts = datetime.fromisoformat(rfc3339.replace("Z", "+00:00"))
    except ValueError:
        return rfc3339
    out = ts + timedelta(minutes=minutes)
    # Preserve the original input's offset rendering style.
    if rfc3339.endswith("Z"):
        return out.astimezone(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
    return out.isoformat()
